drivers/i3c/i3c_npcx.c

```diff
@@ -0,0 +1,2047 @@
+/*
+ * Copyright (c) 2024 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/*132*/
+#define DT_DRV_COMPAT nuvoton_npcx_i3c
+
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/irq.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/sys_io.h>
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/i3c.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/reset.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(npcx_i3c, CONFIG_I3C_LOG_LEVEL);
+
+#define NPCX_I3C_CHK_TIMEOUT 10000    /* Unit: us, timeout for checking register status */
+#define I3C_SCL_PP_FREQ_MAX  12500000 /* Unit: Mhz*/
+#define I3C_SCL_OD_FREQ_MAX  4170000  /* Unit: Mhz*/
+
+#define I3C_BUS_TLOW_PP_MIN_NS  24  /* T_LOW period in push-pull mode */
+#define I3C_BUS_THigh_PP_MIN_NS 24  /* T_High period in push-pull mode */
+#define I3C_BUS_TLOW_OD_MIN_NS  200 /* T_LOW period in open-drain mode */
+
+#define PPBAUD_DIV_MAX (BIT(GET_FIELD_SZ(NPCX_I3C_MCONFIG_PPBAUD)) - 1) /* PPBAUD divider max */
+
+#define DAA_TGT_INFO_SZ 0x8 /* 8 bytes = PID(6) + BCR(1) + DCR(1) */
+#define BAMATCH_DIV     0x4 /* BAMATCH = APB4_CLK divided by four */
+
+/* Default maximum time we allow for an I3C transfer (unit:ms) */
+#define I3C_TRANS_TIMEOUT K_MSEC(100)
+
+/* Supported I3C MCLKD frequency */
+enum npcx_i3c_speed {
+	NPCX_I3C_BUS_SPEED_45MHZ,
+};
+
+/* Operation type */
+enum npcx_i3c_oper_state {
+	NPCX_I3C_IDLE,
+	NPCX_I3C_WR,
+	NPCX_I3C_RD,
+};
+
+/* I3C timing configuration for each i3c speed */
+struct npcx_i3c_timing_cfg {
+	uint8_t ppbaud; /* Push-Pull high period */
+	uint8_t pplow;  /* Push-Pull low period */
+	uint8_t odhpp;  /* Open-Drain high period */
+	uint8_t odbaud; /* Open-Drain low period */
+};
+
+/* Recommended I3C timing values are based on MCLKD 45 MHz */
+static const struct npcx_i3c_timing_cfg npcx_def_speed_cfg[] = {
+	/* PP = 12.5 mhz, OD = 4.17 Mhz */
+	[NPCX_I3C_BUS_SPEED_45MHZ] = {.ppbaud = 1, .pplow = 0, .odhpp = 1, .odbaud = 4},
+};
+
+struct npcx_i3c_config {
+	/* Common I3C Driver Config */
+	struct i3c_driver_config common;
+
+	/* Pointer to controller registers. */
+	struct i3c_reg *base;
+
+	/* Pointer to the clock device. */
+	const struct device *clock_dev;
+
+	/* Reset controller */
+	struct reset_dt_spec reset;
+
+	/* Clock control subsys related struct. */
+	struct npcx_clk_cfg clock_subsys;
+
+	/* Reference clock to determine 1 μs bus available time */
+	struct npcx_clk_cfg ref_clk_subsys;
+
+	/* Pointer to pin control device. */
+	const struct pinctrl_dev_config *pincfg;
+
+	/* Interrupt configuration function. */
+	void (*irq_config_func)(const struct device *dev);
+
+#ifdef CONFIG_I3C_NPCX_DMA
+	struct npcx_clk_cfg mdma_clk_subsys;
+	struct mdma_reg *mdma_base;
+#endif
+};
+
+struct npcx_i3c_data {
+	struct i3c_driver_data common;       /* Common i3c driver data */
+	struct k_sem lock_sem;               /* Mutex of i3c controller */
```

Comment:
shouldn't this be a 'mutex' to prevent a priority inversion

Reply:
Done, thanks